src/robusta/integrations/kubernetes/custom_models.py

```diff
@@ -517,6 +525,21 @@ def run_simple_job_spec(
             pod = job.get_single_pod()
             return pod.get_logs() or ""
         finally:
+            if pod and finalizers:
+                try:  # must use patch, since the pod revision changed at this point
+                    pod_finalizers = pod.metadata.finalizers
+                    for finalizer in finalizers:
+                        if finalizer in pod_finalizers:
+                            pod_finalizers.remove(finalizer)
+
+                    client.CoreV1Api().patch_namespaced_pod(
```

Comment:
There is a race condition here if another controller put finalizers and removes them while this runs.

Is it possibleto do a PATCH w/ `op: remove` and explicitly remove only the finalizers that we set w/o impacting others?

Reply:
I don't know if you can do that in a patch
do you?

Reply:
Yes, the syntax is a little obscure and poorly documented, but you can do
something like:

```
patch = {
     "metadata": {
         "$deleteFromPrimitiveList/finalizers": ["finalizer.example.com/foo"]
     }
}
```

Here is a full test case:

```
from kubernetes import client, config
from kubernetes.client.rest import ApiException

def main():
    config.load_kube_config()
    api_instance = client.CoreV1Api()

    pod_manifest = {
        "apiVersion": "v1",
        "kind": "Pod",
        "metadata": {
            "name": "example-pod",
            "finalizers": ["finalizer1.example.com/bar", "
finalizer2.example.com/foo"]
        },
        "spec": {
            "containers": [{
                "name": "nginx",
                "image": "nginx:1.19.0"
            }]
        }
    }

    try:
        api_instance.create_namespaced_pod(namespace="default",
body=pod_manifest)
        print("Pod created with two finalizers.")
    except ApiException as e:
        print(f"Exception when creating a pod: {e}\n")

    patch = {
            "metadata": {
                 "$deleteFromPrimitiveList/finalizers": ["finalizer2.example.com/foo"]
            }
        }
    try:
        x=api_instance.patch_namespaced_pod(name="example-pod", namespace="default", body=patch)
        print("One finalizer removed from pod.")
        print(x)
    except ApiException as e:
        print(f"Exception when patching the pod to remove a finalizer:
{e}\n")

if __name__ == '__main__':
    main()
```

Reply:
are you sure this is a part of the API and not something that will break in future versions?

Reply:
It is documented [here](https://github.com/kubernetes/community/blob/2f8415788e5a50dec90383408185b535b0e55b65/contributors/devel/sig-api-machinery/strategic-merge-patch.md?plain=1#L177) and is the idiomatic way (as far as I can tell) to remove finalizers. Here is an [official code sampe](https://github.com/kubernetes-client/java/blob/fd530a9126e0ffe80d918ccd7b49b00aa4dde382/examples/examples-release-20/src/main/java/io/kubernetes/client/examples/PatchExample.java#L47) and [here is code from the mysql operator using it](https://github.com/mysql/mysql-operator/blob/4a80d27486c36a1ba1262d79cddbb99be21e52ba/mysqloperator/controller/innodbcluster/cluster_api.py#L2330). It is also widely used internally in the Kubernetes codebase itself.

So I think it is safe to use and better than risking the race condition.

Reply:
I don't think there's any real chance of race condition
but it's it's officially supported, it's a better way to do it